working/static metaprogramming/intro.md

```diff
@@ -0,0 +1,431 @@
+at a distance.
+
+## Non-goals
+
+* **Modifying existing code**. We don’t see a need to allow modifiying existing
+  code through metaprogramming. This leads to unexpected behavior, and
+  negatively affects usability. Dart code that already has clearly defined
+  meaning today should continue to mean the same thing after metaprogramming.
+  The goal is not to be able to hijack existing Dart syntax to do wildly
+  different things, but to augment code with new capabilities.
+
+## Approach
+
+To understand what a static metaprogramming solution might look like, we’ll walk
+through one of the use cases. Say a Dart user wants to write a library that
+automatically defines `copyWith()` methods for classes. For them to do that,
+Dart needs to provide a way for users to:
+
+1. **Invoke the metaprogramming** to control which classes should get the
+   `copyWith()` method.
+2. **Define some metaprogramming** that specifies how the copyWith() method is
+   defined in terms of the class’s fields. That metaprogramming needs to be
+   authored in some kind of metaprogramming language.
+3. It uses some sort of **introspection mechanism** to walk over the class’s
+   field declarations.
+4. Based on that it uses some other **construction mechanism** to produce a new
+   method including its signature and body. The resulting is then added to the
+   original class declaration.
+
+The approach we propose is **macros**, and the above requirements list can be
+stated more concretely as:
+
+1. We design some sort of syntax to apply a macro by name to a given class,
+   method declaration, etc. We haven’t spent time here yet.
+2. Metaprogramming happens by defining a macro, which can be called at compile
+   time.
+3. The introspection system is TBD. It could be an imperative API similar to the
+   existing mirrors API or analyzer API (but run at compile-time), or perhaps
+   something declarative like to pattern matching over syntax examples.
+4. We could use an imperative API to construct new pieces of Dart code in a
+   macro, but some sort of quotation syntax would make this nicer and make the
+   API less brittle.
+
+Most of this is as-yet-undesigned, but the key piece is that **we think macros
+should be written in normal imperative Dart code which is executed at compile
+time**.
+
+## Compile-time Dart execution
+
+Allowing a general-purpose, Turing-complete imperative language to run at
+compile time is a big ask. Here’s why we think it’s a good choice:
+
+### We already do it
+
+A subset of the Dart language is already allowed at compile time, which is
+specifically carved out to support limited const expressions. This means users
+are familiar with the concept of “running” Dart code at compile time, and we
+have invested some implementation time in supporting it. You could think of this
+proposal as merely (massively) expanding the subset of Dart allowed in a const
+expression.
```

Comment:
There are separate proposals to do just that, are we considering them part of this, a layer below (i.e. prerequisite to) this, or orthogonal to this?

Reply:
The proposal to expand const is effectively a prerequisite to this proposal. That will essentially lay the groundwork for how we actually run user code at compile time, what rules are in place for that, etc.